Configurator: pass linking opts to the C linker when appropriate

AltGr opened the pull request:
Yet another PR related to my attempts at cross-compilation.

Configurator retrieves the output of `pkg-config --libs` and passes
that to the OCaml linker. These options are generally of the form
`-lfoo`, so this works, but you might encounter different ones that
`ocamlc` won't be able to cope with¹.

This patch only passes `-l*` arguments to OCaml directly, and
redirects anything else to the C linker by prefixing it with `-ccopt`
in the ocaml command-line.

Note: a similar issue might happen with `--cflags` too ? I didn't
encounter it, but we might want to be consistent and only accept
`-I`/`-l` options there as well ?


¹ I am not sure how often that happens of if it should, but the
installed packages can actually put anything in their .pc files, if I
understand correctly.

rgrinberg commented:
Thanks for the patch @AltGr. Could you please sign-off your commit according to the DCO? This is a requirement from all contributors.

I see that CI is currently failing: https://travis-ci.org/ocaml/dune/jobs/595206365 In this particular case, it seems like you've accidentally changed the evaluation order for `pkg-config --libs` and `pkg-config --cflags`. It probably doesn't matter, but let's make this evaluation order explicit and maintain the old one to avoid extra changes to the test suite.

Make sure to run the tests (`$ make test`) and promote the result (`$ make promote`) if you find it satisfactory. Run `$ make format` to format your changes as well btw.

> Note: a similar issue might happen with --cflags too ? I didn't
encounter it, but we might want to be consistent and only accept
-I/-l options there as well ?

I agree with this assessment.

Do you have an example of a package that reproduces this problem? I think it would be useful to add a test case for this issue. We don't have to depend on a real C library in the dune tests, but I think we can just copy some .pc files and then test how the flags and ocamlc invocation that configurator spits out.

AltGr commented:
Fixed the evaluation order and added some tests (based on the MXE .pc file for gtk3 and lablgtk3).

Note that I was slightly incorrect when saying that dune used the results of `pkg-config`: in fact it is the package rules which retrieve these flags, generate sexp files, and then make dune use them. So we could delegate these rules to the packages ; I still think they are more convenient here though.

AltGr commented:
Nevermind, this doesn't seem to be needed anymore on dune master, and actually hurts some C compilation checks. I think it can be closed.
